### DISCUSSION TITLE
Can I display images stored dynamically in next using app directory?

### DISCUSSION THREAD

**Discussion by gabriel-logan** [vercel/next.js#64954] [App Router]
I'm developing an application, in which the user will have a profile photo, I did all the code and everything works ok, stores it correctly in the public/images/ETCETECET.png folder

In dev mode everything works ok.

but when I build and then run yarn start, if the user sends a new photo, I can't see it, I wanted to know if this is really impossible to do DIRECTLY in nextjs.

**Comment by icyJoseph** [comment]
Yeah, you can't use the public folder to store stuff at runtime.

- https://github.com/vercel/next.js/discussions/15089#discussioncomment-36357
- https://github.com/orgs/vercel/discussions/241#discussioncomment-363002
- https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

**Reply by gabriel-logan** [reply]
So I can't serve images statically like express does natively.

Can you tell me if I can use express inside next then? Or if I can use any other server within nextjs, using the same IP

**Reply by icyJoseph** [reply]
How are you hosting this?

**Reply by gabriel-logan** [reply]
I didn't quite understand the question, I'm still developing the system, I'm doing it locally, my idea is more or less the ideas below

![image](https://github.com/vercel/next.js/assets/96627244/0e0a7732-9424-42e4-a3bf-7a0d58f2be78)

![image](https://github.com/vercel/next.js/assets/96627244/862e2530-6a9d-4078-9aa9-f9d8b2c61e32)

![image](https://github.com/vercel/next.js/assets/96627244/75a4a4ce-d208-4b4a-95e9-cb3a78bb0bf0)

With express I was able to serve, you can say, oh you can use external services, for financial reasons, I was choosing languages and versions in which I would spend less resources, so in this case express was useful as I could statically serve images files etc directly on the server itself and just store the information in mysql, then I discovered next

I saw that you could make a web page linked directly to the server and that seemed great, except for the unfortunate part that you can't serve images, I also thought about storing the base64 in the database, but that would be very bad performance, if you could help me Giving an insight would be very helpful.

The initial real project will be with Nextjs (IF IT WORKS, I'm still testing), react native, mysql OR postgress and will also have websockets.

**Reply by gabriel-logan** [reply]
I forgot to mention but everything is being saved perfectly in MySQL

![image](https://github.com/vercel/next.js/assets/96627244/1db30334-8a70-43c7-85c9-f6dd90735dd0)

**Reply by icyJoseph** [reply]
I mean, what you need is a proper way to store, and retrieve the assets. If you host in Vercel or any other serverless oriented host, you need a - 3rd party object storage. Because serverless deploys are taken down after some time, and they don't allow writing to file-system. That's gonna be the issue.

If you host this "yourself" then you can write to file-system, or anything you like, and have incoming requests read from the folder where you save, sort of making your own static server. You'd need a Route Handler for this, rather than the public asset folder. 

Additionally, the Next.js router, iirc, creates a manifest with all allowed public assets, at build time, and that can't be changed at run-time, so requests to those assets just see a 404. 

I'd say is almost easier to spin up a express server, that handles all or this, as a separate server. Next.js is more of a backend for the frontend framework, rather than a fully fledged backend server.

**Reply by gabriel-logan** [reply]
Ok, these tips help a lot, at first, I'm going to deploy on my own server using nginx with reverse proxy, if you have any tips on this too, I'd appreciate it in advance. Thank you.

**Comment by sensorario** [comment]
Can this help you:
https://nextjs.org/docs/pages/building-your-application/optimizing/static-assets
